core/ir/isa_regdeps/decode.c

```diff
@@ -0,0 +1,130 @@
+ */
+byte *
+decode_isa_regdeps(dcontext_t *dcontext, byte *encoded_instr, instr_t *instr)
+{
+    /* Interpret the first 4 bytes of encoded_instr (which are always present) as a uint
+     * for easier retrieving of category, eflags, #src, and #dst values.
+     * We can do this safely because encoded_instr is 4 bytes aligned.
+     */
+    ASSERT(ALIGNED(encoded_instr, ALIGN_BYTES));
+    uint encoding_header = *((uint *)encoded_instr);
+
+    /* Decode number of register destination operands.
+     */
+    uint num_dsts = encoding_header & DST_OPND_MASK;
+
+    /* Decode number of register source operands.
+     */
+    uint num_srcs = (encoding_header & SRC_OPND_MASK) >> SRC_OPND_SHIFT;
+
+    instr_set_num_opnds(dcontext, instr, num_dsts, num_srcs);
+
+    /* Decode arithmetic flags.
+     */
+    uint eflags = (encoding_header & FLAGS_MASK) >> FLAGS_SHIFT;
+    uint eflags_instr = 0;
+    if (TESTANY(REGDEPS_INSTR_WRITES_ARITH, eflags))
+        eflags_instr |= EFLAGS_WRITE_ARITH;
+    if (TESTANY(REGDEPS_INSTR_READS_ARITH, eflags))
+        eflags_instr |= EFLAGS_READ_ARITH;
+    instr->eflags = eflags_instr;
+
+    /* Declare the eflags to be valid.
+     * This is needed in order to retrieve their value without trying to compute it again.
+     */
+    instr_set_arith_flags_valid(instr, true);
+
+    /* Decode instruction category.
+     */
+    uint category = (encoding_header & CATEGORY_MASK) >> CATEGORY_SHIFT;
+    instr_set_category(instr, category);
+
+    /* Decode operation size, if there are any operands.
+     */
+    uint num_opnds = num_dsts + num_srcs;
+    opnd_size_t max_opnd_size = OPSZ_0;
+    if (num_opnds > 0)
+        max_opnd_size = (opnd_size_t)encoded_instr[OP_SIZE_INDEX];
+    instr->operation_size = max_opnd_size;
+
+    /* Decode register destination operands, if present.
+     */
+    for (uint i = 0; i < num_dsts; ++i) {
+        reg_id_t dst = (reg_id_t)encoded_instr[i + OPND_INDEX];
+        opnd_t dst_opnd = opnd_create_reg((reg_id_t)dst);
+        instr_set_dst(instr, i, dst_opnd);
+    }
+
+    /* Decode register source operands, if present.
+     */
+    for (uint i = 0; i < num_srcs; ++i) {
+        reg_id_t src = (reg_id_t)encoded_instr[i + OPND_INDEX + num_dsts];
+        opnd_t src_opnd = opnd_create_reg((reg_id_t)src);
+        instr_set_src(instr, i, src_opnd);
+    }
+
+    /* Declare the operands to be valid.
+     */
+    instr_set_operands_valid(instr, true);
+
+    /* Compute instruction length including bytes for padding to reach 4 bytes alignment.
+     * Account for 1 additional byte containing max register operand size, if there are
+     * any operands.
+     */
+    uint num_opnd_bytes = num_opnds > 0 ? num_opnds + 1 : 0;
+    uint instr_length = ALIGN_FORWARD(HEADER_BYTES + num_opnd_bytes, ALIGN_BYTES);
+
+    /* Set decoded instruction ISA mode to be synthetic.
+     */
+    instr_set_isa_mode(instr, DR_ISA_REGDEPS);
+
```

Comment:
The raw bits should be set to valid so an encoding or length query will not try to re-encode.

Reply:
Done. instr_allocate_raw_bits() sets the raw bits to valid (and allocated).